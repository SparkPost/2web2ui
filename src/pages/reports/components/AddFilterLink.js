import React, { useState } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import qs from 'qs';

import { addFilters } from 'src/actions/reportOptions';
import { PageLink } from 'src/components/links';
import { stringifyTypeaheadfilter } from 'src/helpers/string';
import {
  selectReportSearchOptions,
  selectSummaryChartSearchOptions,
} from 'src/selectors/reportSearchOptions';

export const AddFilterLink = ({
  //From parent
  newFilter,
  reportType,
  content,
  //From redux
  currentSearchOptions,
  addFilters,
}) => {
  const [isNewTabKeyPressed, setNewTabKeyPressed] = useState(false);

  const currentFilters = currentSearchOptions.filters || [];

  /**
   * Needs to first check if cmd/ctrl key is pressed when mouse is released
   * Then it's saved for use in the onClick handler. We need both because
   * onClick does not have the correct values for metaKey and ctrlKey
   **/
  const handleMouseUp = e => {
    setNewTabKeyPressed(e.metaKey || e.ctrlKey);
  };

  const handleClick = () => {
    if (!isNewTabKeyPressed) {
      addFilters([newFilter]);
    }
  };

  const mergedFilters = _.uniqWith(
    [...currentFilters, stringifyTypeaheadfilter(newFilter)],
    _.isEqual,
  );
  const newSearchOptions = { ...currentSearchOptions, filters: mergedFilters };
  const linkParams = qs.stringify(newSearchOptions, { arrayFormat: 'repeat' });
  const fullLink = `/reports/${reportType}/?${linkParams}`;

  return (
    <PageLink onMouseUp={handleMouseUp} onClick={handleClick} to={fullLink} replace>
      {content}
    </PageLink>
  );
};

const mapStateToProps = (state, props) => ({
  currentSearchOptions:
    props.reportType === 'summary'
      ? selectSummaryChartSearchOptions(state)
      : selectReportSearchOptions(state),
});

export default connect(mapStateToProps, { addFilters })(AddFilterLink);
