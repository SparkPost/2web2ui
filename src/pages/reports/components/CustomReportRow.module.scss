@import '~@sparkpost/matchbox/src/styles/config.scss';

.Row {
  display: flex;
  border-bottom: 1px solid color(gray, 9);

  &:last-child {
    border-bottom: none;
  }

  &:hover {
    cursor: pointer;
  }
}

.Delete {
  flex: 0 0;
  align-self: flex-end;
  padding: spacing();
}

.DeleteButton {
  color: color(gray, 4);

  &:hover {
    color: color(red)
  }
}

.Name {
  flex: 1 0;
  padding: spacing();
  line-height: line-height(600);
  font-size: font-size(500);

  color: color(blue, dark);

  &:hover {
    text-decoration: underline;
  }
}
