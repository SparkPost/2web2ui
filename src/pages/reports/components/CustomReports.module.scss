@import '~@sparkpost/matchbox/src/styles/config.scss';

.NameWrapper {
  input {
    font-size: font-size(500);
  }
}

.Wrap {
  display: flex;
  align-items: baseline;

  & > *:not(:first-child) {
    margin-left: spacing();
  }
}

.ReportsButton {
  flex: 1 0;
  text-align: right;
}

.SaveIcon {
  margin-top: -0.2em;
  margin-right: 0.5ch;
}

.Limited {
  font-size: font-size(200);
  line-height: line-height(500);
  color: color(gray, 5);
}
