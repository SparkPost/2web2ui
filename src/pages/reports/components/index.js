export { default as ReportOptions } from './ReportOptions';
export { default as MetricCard } from './MetricCard';
export { default as MetricsSummary } from './MetricsSummary';
export { default as ShareModal } from './ShareModal';
