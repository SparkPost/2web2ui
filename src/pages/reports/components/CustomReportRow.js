import React from 'react';
import { Button } from 'src/components/matchbox';
import styles from './CustomReportRow.module.scss';

function ReportRow(props) {
  return (
    <div onClick={props.onLoad} className={styles.Row}>
      <div className={styles.Name}>{props.name}</div>
      <div className={styles.Delete}>
        <Button
          className={styles.DeleteButton}
          disabled={props.deleting}
          flat
          size="small"
          onClick={e => {
            e.stopPropagation();
            props.onDelete(props.name);
          }}
        >
          Delete
        </Button>
      </div>
    </div>
  );
}

export default ReportRow;
