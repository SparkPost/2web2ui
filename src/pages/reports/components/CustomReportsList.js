import React from 'react';
import { withRouter } from 'react-router-dom';
import { Button, Panel, Modal } from 'src/components/matchbox';
import ReportRow from './CustomReportRow';

export function CustomReportsList({ deleting, handleLoad, onDelete, reports }) {
  const [open, setOpen] = React.useState(false);

  const reportList = React.useMemo(() => {
    if (!reports.length) {
      return <Panel.LEGACY.Section>You do not have any saved reports.</Panel.LEGACY.Section>;
    }

    return reports.map((report, i) => {
      function onLoad() {
        setOpen(false);
        handleLoad(report);
      }

      return (
        <ReportRow
          key={`${report.name}-${i}`}
          {...report}
          onLoad={onLoad}
          onDelete={onDelete}
          deleting={deleting}
        />
      );
    });
  }, [deleting, handleLoad, onDelete, reports]);

  return (
    <>
      <Button flat color="orange" size="small" onClick={() => setOpen(!open)}>
        Saved Reports
      </Button>
      <Modal.LEGACY open={open} onClose={() => setOpen(false)} showCloseButton>
        <Panel.LEGACY title="Saved Reports">{reportList}</Panel.LEGACY>
      </Modal.LEGACY>
    </>
  );
}

export default withRouter(CustomReportsList);
