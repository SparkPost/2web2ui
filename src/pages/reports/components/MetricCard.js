import React from 'react';
import PropTypes from 'prop-types';
import { InfoOutline } from '@sparkpost/matchbox-icons';
import useHibanaOverride from 'src/hooks/useHibanaOverride';
import useUniqueId from 'src/hooks/useUniqueId';
import { Panel, Tooltip } from 'src/components/matchbox';
import { Heading } from 'src/components/text';
import OGStyles from './MetricCard.module.scss';
import hibanaStyles from './MetricCardHibana.module.scss';

const MetricCard = ({ label, value, tooltipContent, ...rest }) => {
  const styles = useHibanaOverride(OGStyles, hibanaStyles);
  const tooltipId = useUniqueId('metric-card');
  const tooltip = tooltipContent ? (
    <Tooltip id={tooltipId} dark content={tooltipContent} horizontalOffset="-1.1rem">
      <InfoOutline className={styles.TooltipIcon} />
    </Tooltip>
  ) : null;

  return (
    <div className={styles.MetricCard}>
      <Panel.LEGACY className={styles.Panel} {...rest}>
        <Panel.LEGACY.Section className={styles.PanelSection}>
          <Heading as="h3" className={styles.Value}>
            {value}
          </Heading>
          <Heading as="h4" className={styles.Label}>
            {label} {tooltip}
          </Heading>
        </Panel.LEGACY.Section>
      </Panel.LEGACY>
    </div>
  );
};

MetricCard.propTypes = {
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.node]),
  tooltipContent: PropTypes.string,
  label: PropTypes.string,
};

export default MetricCard;
