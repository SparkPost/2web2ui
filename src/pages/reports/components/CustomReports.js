import React from 'react';
import { connect } from 'react-redux';
import { addFilters, clearFilters, refreshReportOptions } from 'src/actions/reportOptions';
import { selectCustomReports } from 'src/selectors/customReports';
import { Save } from '@sparkpost/matchbox-icons';
import { saveReport, deleteReport } from 'src/actions/customReports';
import { withRouter } from 'react-router-dom';
import { Button, Panel, TextField } from 'src/components/matchbox';
import CustomReportsList from './CustomReportsList';
import { parseSearch } from 'src/helpers/reports';
import qs from 'query-string';
import _ from 'lodash';

import styles from './CustomReports.module.scss';

const REPORT_LIMIT = 10;

export function CustomReports(props) {
  const {
    addFilters,
    deleteReport,
    clearFilters,
    searchOptions,
    loading,
    refreshReportOptions,
    reports,
    saveReport,
  } = props;

  const [name, setName] = React.useState('');

  function handleNameChange(e) {
    const value = e.target.value;

    // Limit string length
    if (value.length < 24) {
      setName(e.target.value);
    }
  }

  function handleSave() {
    let toSave = searchOptions;

    if (searchOptions.range !== 'custom') {
      toSave = _.omit(toSave, ['to', 'from']);
    }

    saveReport({
      name: name.trim(),
      url: qs.stringify(toSave),
    });
  }

  function handleLoad(report) {
    const { options, filters = [] } = parseSearch(report.url);
    clearFilters();
    addFilters(filters);
    refreshReportOptions(options);
    setName(report.name);
  }

  const isLimited = reports.length >= REPORT_LIMIT;
  const canOverwrite = _.find(reports, ['name', name.trim()]);
  const saveDisabled = name === '' || loading;

  return (
    <Panel.LEGACY.Section>
      <div className={styles.Wrap}>
        <div className={styles.NameWrapper}>
          <TextField
            id="custom-report-name"
            label="Report Name"
            labelHidden
            placeholder="Untitled Report"
            onChange={handleNameChange}
            value={name}
          />
        </div>

        {!isLimited || (canOverwrite && isLimited) ? (
          <Button disabled={saveDisabled} onClick={handleSave} flat color="orange">
            <Save className={styles.SaveIcon} />
            {loading ? 'Saving...' : 'Save'}
          </Button>
        ) : null}

        {isLimited && (
          <div className={styles.Limited}>You are limited to {REPORT_LIMIT} saved reports</div>
        )}

        <div className={styles.ReportsButton}>
          <CustomReportsList
            deleting={loading}
            handleLoad={handleLoad}
            reports={reports}
            onDelete={deleteReport}
          />
        </div>
      </div>
    </Panel.LEGACY.Section>
  );
}

const mapStateToProps = state => ({
  reports: selectCustomReports(state),
  loading: state.currentUser.userOptionsPending,
});

const mapDispatchToProps = {
  addFilters,
  clearFilters,
  refreshReportOptions,
  saveReport,
  deleteReport,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(CustomReports));
