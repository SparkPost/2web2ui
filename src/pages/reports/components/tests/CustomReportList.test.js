import React from 'react';
import { shallow } from 'enzyme';
import { Modal } from 'src/components/matchbox';
import { CustomReportsList } from '../CustomReportsList';

describe('Component: CustomReportsList', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      reports: [
        { name: 'report 1', url: 'report1url' },
        { name: 'report 2', url: 'report2url' },
      ],
      handleLoad: jest.fn(),
      onDelete: jest.fn(),
      deleting: false,
    };
    wrapper = shallow(<CustomReportsList {...props} />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render empty state with no reports', () => {
    wrapper.setProps({ reports: [] });
    expect(wrapper.contains('You do not have any saved reports.')).toBe(true);
  });

  it('should open the modal when clicking saved reports', () => {
    wrapper.find('Button').simulate('click');
    expect(wrapper.find(Modal.LEGACY).prop('open')).toBe(true);
  });

  it('should close the modal with the onClose callback', () => {
    wrapper.find('Button').simulate('click');
    wrapper.find(Modal.LEGACY).simulate('close');
    expect(wrapper.find(Modal.LEGACY).prop('open')).toBe(false);
  });

  it('should load a report with the onLoad callback', () => {
    wrapper.find('Button').simulate('click');
    wrapper
      .find('ReportRow')
      .first()
      .simulate('load');
    expect(props.handleLoad).toHaveBeenCalledWith(props.reports[0]);
    expect(wrapper.find(Modal.LEGACY).prop('open')).toBe(false);
  });
});
