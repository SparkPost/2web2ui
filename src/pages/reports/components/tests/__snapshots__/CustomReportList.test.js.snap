// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CustomReportsList should render 1`] = `
<Fragment>
  <Button
    color="orange"
    flat={true}
    onClick={[Function]}
    size="small"
  >
    Saved Reports
  </Button>
  <Modal.LEGACY
    onClose={[Function]}
    open={false}
    showCloseButton={true}
  >
    <Panel.LEGACY
      title="Saved Reports"
    >
      <ReportRow
        deleting={false}
        key="report 1-0"
        name="report 1"
        onDelete={[MockFunction]}
        onLoad={[Function]}
        url="report1url"
      />
      <ReportRow
        deleting={false}
        key="report 2-1"
        name="report 2"
        onDelete={[MockFunction]}
        onLoad={[Function]}
        url="report2url"
      />
    </Panel.LEGACY>
  </Modal.LEGACY>
</Fragment>
`;
